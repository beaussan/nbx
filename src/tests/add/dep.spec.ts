import { filesystem } from 'gluegun';
import * as latestVersion from 'latest-version';
import Dep from '../../commands/add/dep';
import { expectFailCli, expectGitCommits, testCli } from '../test-helpers.spec';

jest.setTimeout(90000);

testCli({
  runCommand: (args: string[]) => Dep.run(args),
  name: 'nbx add:dep',
  defaultArgs: ['--no-spinner'],
  tests: [
    {
      name: 'fail if no package.json found',
      runner: expectFailCli({
        args: ['chalk'],
        errorMessage: 'There is no package.json not found in the current folder',
        before: async () => {
          filesystem.remove('package.json');
        },
      }),
    },
    /*
     */
    {
      name: 'fail if dependancy is already in package.json',
      runner: expectFailCli({
        args: ['chalk'],
        errorMessage: 'chalk is already installed in this project.',
        packageJson: { dependencies: { chalk: '2.3.2' } },
        withGitSetup: true,
      }),
    },

    {
      name: 'fail if dev dependancy is already in package.json',
      runner: expectFailCli({
        args: ['chalk', '-D'],
        errorMessage: 'chalk is already installed in this project.',
        packageJson: { devDependencies: { chalk: '2.3.2' } },
        withGitSetup: true,
      }),
    },
    {
      name: 'work with dependency',
      runner: expectGitCommits({
        args: ['chalk'],
        expectGitLog: async () => {
          const latestChalk = await latestVersion('chalk');
          return [
            `:heavy_plus_sign: add chalk@${latestChalk} as a dependency`,
            'M\tpackage.json',
            'M\tyarn.lock',
          ];
        },
        checks: async () => {
          const latestChalk = await latestVersion('chalk');

          const packageJson = filesystem.read('package.json', 'json');
          expect(packageJson).toStrictEqual({
            dependencies: {
              chalk: latestChalk,
            },
          });
        },
      }),
    },
    {
      name: 'work with a dev dependency',
      runner: expectGitCommits({
        args: ['chalk', '-D'],
        expectGitLog: async () => {
          const latestChalk = await latestVersion('chalk');
          return [
            `:heavy_plus_sign: add chalk@${latestChalk} as a dev dependency`,
            'M\tpackage.json',
            'M\tyarn.lock',
          ];
        },
        checks: async () => {
          const latestChalk = await latestVersion('chalk');

          const packageJson = filesystem.read('package.json', 'json');
          expect(packageJson).toStrictEqual({
            devDependencies: {
              chalk: latestChalk,
            },
          });
        },
      }),
    },
  ],
});
