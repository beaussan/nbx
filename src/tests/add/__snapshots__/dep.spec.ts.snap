// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nbx add:dep should fail if dependancy is already in package.json 1`] = `
Array [
  "Adding chalk as a dependency",
  "The step was done without any error.",
]
`;

exports[`nbx add:dep should fail if dev dependancy is already in package.json 1`] = `Array []`;

exports[`nbx add:dep should fail if no package.json found 1`] = `Array []`;

exports[`nbx add:dep should print help correctly 1`] = `
Array [
  "proxy to yarn add, plus gitmoji commit
",
  "USAGE
  $ nbx add:dep DEP

ARGUMENTS
  DEP  The dependency to install

OPTIONS
  -D, --dev       install as a dev dependency
  -h, --help      show CLI help
  -v, --verbose   Verbose output
  --[no-]spinner  Enable spinner in cli output, true by default

EXAMPLES
  $ nbx add:dep -D eslint
  $ nbx add:dep --dev eslint
  $ nbx add:dep chalk",
  "",
]
`;

exports[`nbx add:dep should work with a dev dependency 1`] = `
Array [
  "Adding chalk as a dev dependency",
  "The step was done without any error.",
]
`;

exports[`nbx add:dep should work with dependency 1`] = `
Array [
  "Adding chalk as a dependency",
  "The step was done without any error.",
]
`;
